File: agent-governance-python/agent-os/src/agent_os/policies/context_aggregation.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.
"""Aggregation evaluation for Context Accumulation Governance.

Applies organization-authored rules over *combinations* of accumulated labels,
plus a monotone backstop so that combinations not covered by a rule escalate for review
rather than passing silently. The framework supplies the evaluation hook; the
specific combination rules are authored per organization.

Aggregation (a collection of declared labels crossing a threshold) is distinct
from inference (semantic derivation of new sensitivity); this module governs
the former. Inference detection is out of scope for v1.
"""

from __future__ import annotations

from dataclasses import dataclass

from .context_envelope import ContextEnvelope
from .data_classification import DataClassification


@dataclass(frozen=True)
class AggregationRule:
    """One organization-authored rule over a label combination.

    When every label in ``all_labels`` is present in an envelope, the rule
    raises sensitivity to at least ``sets_sensitivity`` and adds
    ``adds_restrictions``.
    """

    name: str
    all_labels: frozenset[str]
    sets_sensitivity: DataClassification
    adds_restrictions: frozenset[str] = frozenset()


@dataclass(frozen=True)
class AggregationRuleSet:
    """An ordered collection of aggregation rules."""

    rules: tuple[AggregationRule, ...] = ()


@dataclass(frozen=True)
class AggregationResult:
    """Outcome of evaluating an envelope against a rule set."""

    aggregate_sensitivity: DataClassification
    restrictions: frozenset[str]
    escalate: bool
    rules_applied: tuple[str, ...]


def evaluate_aggregation(
    env: ContextEnvelope,
    ruleset: AggregationRuleSet,
    n_category_threshold: int,
) -> AggregationResult:
    """Evaluate ``env`` against ``ruleset`` and apply the monotone backstop.

    The result sensitivity is the max of the envelope's current sensitivity
    (already the running max of folded per-datum classifications) and every
    matching rule's ``sets_sensitivity`` — so an envelope that grows without
    crossing a declared combination keeps its current classification.

    Escalation fires when no rule governs the envelope yet it has accumulated
    at least ``n_category_threshold`` distinct labels — the conservative
    backstop for combinations not covered by a rule.
    """
    sensitivity = env.aggregate_sensitivity
    restrictions: set[str] = set(env.restrictions)
    applied: list[str] = []

    for rule in ruleset.rules:
        if rule.all_labels <= env.labels:
            sensitivity = max(sensitivity, rule.sets_sensitivity)
            restrictions |= set(rule.adds_restrictions)
            applied.append(rule.name)

    escalate = not applied and len(env.labels) >= n_category_threshold
    return AggregationResult(
        aggregate_sensitivity=sensitivity,
        restrictions=frozenset(restrictions),
        escalate=escalate,
        rules_applied=tuple(applied),
    )

File: agent-governance-dotnet/src/AgentGovernance/Context/ContextAggregation.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

using System.Collections.ObjectModel;

namespace AgentGovernance.Context;

/// <summary>
/// Organization-authored rule over a combination of accumulated labels.
/// </summary>
public sealed class AggregationRule
{
    /// <summary>Creates an immutable aggregation rule.</summary>
    public AggregationRule(
        string name,
        IEnumerable<string> allLabels,
        DataClassification setsSensitivity,
        IEnumerable<string>? addsRestrictions = null)
    {
        Name = ContextTokens.Identifier(name, nameof(name));
        AllLabels = ContextTokens.Snapshot(allLabels, nameof(allLabels));
        SetsSensitivity = DataClassificationGuard.Defined(
            setsSensitivity,
            nameof(setsSensitivity));
        AddsRestrictions = ContextTokens.SnapshotOptional(
            addsRestrictions,
            nameof(addsRestrictions));
    }

    /// <summary>Stable rule name.</summary>
    public string Name { get; }

    /// <summary>Labels that must all be present for this rule to match.</summary>
    public IReadOnlySet<string> AllLabels { get; }

    /// <summary>Minimum sensitivity applied by this rule.</summary>
    public DataClassification SetsSensitivity { get; }

    /// <summary>Restrictions added by this rule.</summary>
    public IReadOnlySet<string> AddsRestrictions { get; }
}

/// <summary>
/// Ordered immutable collection of aggregation rules.
/// </summary>
public sealed class AggregationRuleSet
{
    /// <summary>Creates a rule set in declared evaluation order.</summary>
    public AggregationRuleSet(IEnumerable<AggregationRule>? rules = null)
    {
        var snapshot = rules?.ToArray() ?? [];
        if (snapshot.Any(static rule => rule is null))
        {
            throw new ArgumentException("Rules cannot contain null values.", nameof(rules));
        }

        Rules = new ReadOnlyCollection<AggregationRule>(snapshot);
    }

    /// <summary>Rules in deterministic evaluation order.</summary>
    public IReadOnlyList<AggregationRule> Rules { get; }

    /// <summary>
    /// Applies known rules and escalates unknown combinations at the threshold.
    /// </summary>
    public AggregationResult Evaluate(
        ContextEnvelope envelope,
        int categoryThreshold)
    {
        ArgumentNullException.ThrowIfNull(envelope);
        if (categoryThreshold < 1)
        {
            throw new ArgumentOutOfRangeException(
                nameof(categoryThreshold),
                categoryThreshold,
                "Category threshold must be positive.");
        }

        var sensitivity = envelope.AggregateSensitivity;
        var restrictions = new HashSet<string>(
            envelope.Restrictions,
            StringComparer.Ordinal);
        var applied = new List<string>();

        foreach (var rule in Rules)
        {
            if (!rule.AllLabels.IsSubsetOf(envelope.Labels))
            {
                continue;
            }

            sensitivity = DataClassificationGuard.Maximum(
                sensitivity,
                rule.SetsSensitivity);
            restrictions.UnionWith(rule.AddsRestrictions);
            applied.Add(rule.Name);
        }

        return new AggregationResult(
            sensitivity,
            restrictions,
            escalate: applied.Count == 0 && envelope.Labels.Count >= categoryThreshold,
            applied);
    }
}

/// <summary>
/// Result of evaluating an envelope against aggregation rules.
/// </summary>
public sealed class AggregationResult
{
    internal AggregationResult(
        DataClassification aggregateSensitivity,
        IEnumerable<string> restrictions,
        bool escalate,
        IEnumerable<string> rulesApplied)
    {
        AggregateSensitivity = DataClassificationGuard.Defined(
            aggregateSensitivity,
            nameof(aggregateSensitivity));
        Restrictions = ContextTokens.Snapshot(restrictions, nameof(restrictions));
        Escalate = escalate;
        RulesApplied = new ReadOnlyCollection<string>(rulesApplied.ToArray());
    }

    /// <summary>Resulting maximum sensitivity.</summary>
    public DataClassification AggregateSensitivity { get; }

    /// <summary>Resulting grow-only restrictions.</summary>
    public IReadOnlySet<string> Restrictions { get; }

    /// <summary>Whether the unknown-combination backstop requires review.</summary>
    public bool Escalate { get; }

    /// <summary>Names of matching rules in evaluation order.</summary>
    public IReadOnlyList<string> RulesApplied { get; }
}
